File: lab9x-andreigabor21/gun-shop-ui/src/app/clients/shared/client.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Client} from './client.model';
import {Observable} from 'rxjs';
import {map} from 'rxjs/operators';
import {Address} from './address.model';
import {GunType} from '../../gun-type/shared/gun-type.model';
import {Rental} from '../../rental/shared/rental.model';

// @Injectable({
//   providedIn: 'root'
// })
@Injectable()
export class ClientService {

  private clientsUrl = 'http://localhost:8080/api/clients';
  private rentalsUrl = 'http://localhost:8080/api/rentals';

  constructor(private httpClient: HttpClient) { }

  getClients(): Observable<Client[]> {
    return this.httpClient.get<Client[]>(this.clientsUrl);
  }

  // saveClient(client: Client): Observable<Client>{
  //   console.log(client);
  //   return this.httpClient.post<Client>(this.clientsUrl, client);
  // }

  saveClient(client: Client, address: Address): Observable<Client>{
      console.log(client);
      console.log(JSON.stringify(client));
      console.log(address);
      client.address = address;
      console.log(client);
      return this.httpClient.post<Client>(this.clientsUrl, client);
    }


  getClient(id: number): Observable<Client> {
    return this.getClients()
      .pipe(
        map(clients => clients.find(client => client.id === id))
      );
  }

  updateClient(client: Client): Observable<Client> {
    const dob = client.dateOfBirth;
    console.log(client);
    const date = new Date(dob);
    date.setHours(19);
    const dateJSON = date.toJSON();
    console.log(dateJSON);
    client.dateOfBirth = dateJSON;
    const url = `${this.clientsUrl}/${client.id}`;
    return this.httpClient
      .put<Client>(url, client);
  }

  deleteClient(client: Client): any {
    const url = `${this.clientsUrl}/${client.id}`;
    return this.httpClient
      .delete(url);
  }

  createRental(client: Client, selectedGunType: GunType): Observable<Rental> {
    const clientId = client.id;
    const gunTypeId = selectedGunType.id;
    const price = 100;
    const rental: Rental = {clientId, gunTypeId, price};
    console.log('rental: ', rental);
    return this.httpClient.post<Rental>(this.rentalsUrl, rental);
  }

  findByCity(city: string): Observable<Client[]> {
    const url = `${this.clientsUrl}/search/${city}`;
    return this.httpClient.get<Client[]>(url);
  }
}

File: lab9x-andreigabor21/gun-shop-backend/catalog-core/src/main/java/ro/ubb/catalog/core/service/ClientService.java
package ro.ubb.catalog.core.service;

import ro.ubb.catalog.core.model.Client;
import ro.ubb.catalog.core.model.GunType;
import ro.ubb.catalog.core.model.Rental;

import java.time.LocalDate;
import java.util.List;
import java.util.Set;

public interface ClientService {
    List<Client> getAllClients();

    Client addClient(Client client);

    void deleteClient(Long id);

    Client updateClient(Client client);

    Client getClientById(Long id);

    List<Client> getAllClientsBornBefore(LocalDate date);

    List<Client> getAllClientsByCity(String city);

    Set<Rental> getRentals();

    Rental addRental(Long clientId, Long gunTypeId, Integer price);

    Rental updateRental(Long clientId, Long gunId, Integer price);

    void deleteRental(Long clientId, Long gunTypeId);

    GunType getMostRentedGunType();
}
